bin/install-theos

```diff
@@ -0,0 +1,532 @@
+#!/usr/bin/env bash
+
+# Error codes + association:
+# 1 - Running as root
+# 2 - Unsupported platform
+# 3 - Dependency issue
+# 4 - Unsupported shell
+# 5 - Setting $THEOS failed
+# 6 - Theos clone failed
+# 7 - Toolchain install failed
+# 8 - SDK install failed
+# 9 - Checkra1n '/opt' setup failed
+# 10 - WSL1 fakeroot->fakeroot-tcp failed
+# 11 - Enabling Linux binary compat on FreeBSD failed
+
+set -e
+
+# Pretty print
+special() {
+	printf "\e[0;34m==> \e[1;34mTheos Installer:\e[m %s\n" "$1"
+}
+
+update() {
+	printf "\n\e[0;36m==> \e[1;36m%s\e[m\n" "$1"
+}
+
+common() {
+	printf "\n\e[0;37m==> \e[1;37m%s\e[m\n" "$1"
+}
+
+error() {
+	printf "\e[0;31m==> \e[1;31m%s\e[m\n" "$1"
+}
+
+
+# Root is no bueno
+if [[ $EUID -eq 0 ]]; then
+	error "Theos should NOT be installed with or run as root (su/sudo)!"
+	error "  - Please re-run the installer as a non-root user."
+	exit 1
+fi
+
+
+# Common vars
+PLATFORM=$(uname)
+CSHELL="${SHELL##*/}"
+SHELL_ENV="unknown"
+if [[ "$CSHELL" == "sh" || "$CSHELL" == "bash" ]]; then
+	SHELL_ENV="$HOME/.profile"
+elif [[ "$CSHELL" == "zsh" ]]; then
+	SHELL_ENV="$HOME/.zshenv"
+fi
+# TODO: CI support (non-interactive)
+NONINTERACTIVE=0
+if [[ -z $CI ]]; then
+	NONINTERACTIVE=1
+fi
+
+
+# The work
+get-theos() {
+	# Get Theos
+	update "Checking for Theos install..."
+	if [[ -d "$THEOS" && "$(ls -A "$THEOS")" ]]; then
+		update "Theos appears to already be installed. Checking for updates..."
+		$THEOS/bin/update-theos
+	else
+		update "Theos does not appear to be installed. Cloning now..."
+		git clone --recursive https://github.com/theos/theos.git $THEOS && update "Git clone of Theos was successful!" || (error "Theos git clone command seems to have encountered an error. Please see the log above." && exit 6)
+	fi
+}
+
+get-sdks() {
+	# Get patched sdks
+	update "Checking for patched sdks..."
+	if [[ -d "$THEOS/sdks/" && "$(ls -A "$THEOS/sdks/" | grep sdk)" ]]; then
+		update "SDKs appear to already be installed."
+	else
+		update "SDKS do not appear to be installed. Installing now..."
+		curl -LO https://github.com/theos/sdks/archive/master.zip
+		TMP=$(mktemp -d)
+		unzip master.zip -d $TMP
+		mv $TMP/sdks-master/*.sdk $THEOS/sdks
+		rm -r master.zip $TMP
+
+		if [[ -d "$THEOS/sdks/" && "$(ls -A "$THEOS/sdks/" | grep sdk)" ]]; then
+			update "SDKs successfully installed!"
+		else
+			error "Something appears to have gone wrong. Please try again."
+			exit 8
+		fi
+	fi
+}
+
+darwin() {
+	# Check for Xcode
+	XCODE="/Applications/Xcode.app/"
+	if ! [[ -d "$XCODE" && "$(ls -A "$XCODE")" ]]; then
+		error "Xcode is mandatory. The Command Line Tools package isn't sufficient for Theos to work. Xcode provides toolchains for all Apple platforms."
```

Comment:
Let's link to <https://developer.apple.com/download/applications/> to encourage users to get Xcode from developer.apple.com instead of the MAS.

In the future, it would be really cool if we could automatically install and use [xcodes](https://github.com/RobotsAndPencils/xcodes) with `--experimental-unxip` + aria2 support.

Reply:
Added the download link to the 'missing Xcode' prompt.